api_handlers.go

```diff
@@ -0,0 +1,146 @@
+//
+// Params:
+//     sha: SHA[0:10] of the repo when the tests were executed (or 'latest')
+func apiTestRunsHandler(w http.ResponseWriter, r *http.Request) {
+	runSHA, err := GetRunSHA(r)
+	if err != nil {
+		http.Error(w, "Invalid query params", http.StatusBadRequest)
+		return
+	}
+
+	ctx := appengine.NewContext(r)
+	var browserNames []string
+	browserNames, err = GetBrowserNames()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	var testRuns []TestRun
+	baseQuery := datastore.NewQuery("TestRun").Order("-CreatedAt").Limit(1)
+
+	for _, browserName := range browserNames {
+		var testRunResults []TestRun
+		query := baseQuery.Filter("BrowserName =", browserName)
+		if runSHA != "" && runSHA != "latest" {
+			query = query.Filter("Revision =", runSHA)
+		}
+		if _, err := query.GetAll(ctx, &testRunResults); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		testRuns = append(testRuns, testRunResults...)
+	}
+
+	testRunsBytes, err := json.Marshal(testRuns)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Write(testRunsBytes)
+}
+
+// apiTestRunHandler is responsible for emitting the test-run JSON a specific run,
+// identified by a named browser (platform) at a given SHA.
+//
+// Params:
+//     sha: SHA[0:10] of the repo when the test was executed (or 'latest')
+//     browser: Browser for the run (e.g. 'chrome', 'safari-10')
+func apiTestRunHandler(w http.ResponseWriter, r *http.Request) {
+	runSHA, err := GetRunSHA(r)
+	if err != nil {
+		http.Error(w, "Invalid query params", http.StatusBadRequest)
+		return
+	}
+
+	var browserName string
+	browserName, err = getBrowserParam(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	} else if browserName == "" {
+		http.Error(w, "Invalid 'browser' param", http.StatusBadRequest)
+		return
+	}
+
+	ctx := appengine.NewContext(r)
+
+	query := datastore.
+		NewQuery("TestRun").
+		Order("-CreatedAt").
+		Limit(1).
+		Filter("BrowserName =", browserName)
+	if runSHA != "" && runSHA != "latest" {
+		query = query.Filter("Revision =", runSHA)
+	}
+
+	var testRuns []TestRun
+	if _, err := query.GetAll(ctx, &testRuns); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if len(testRuns) == 0 {
+		http.NotFound(w, r)
+		return
+	}
+
+	testRunsBytes, err := json.Marshal(testRuns[0])
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Write(testRunsBytes)
+}
+
+// getBrowserParam parses and validates the 'browser' param for the request.
+// It returns "" by default (and in error cases).
+func getBrowserParam(r *http.Request) (browser string, err error) {
+	browser = ""
+	params, err := url.ParseQuery(r.URL.RawQuery)
+	if err != nil {
+		return browser, err
```

Comment:
Nit: in some places in this PR the default is returned (naked return), whereas here it's explicit. I don't have a preference other than being consistent.

Reply:
We should aim to be explicit - naked is unreadable IMO

Reply:
Looks like you're referring to methods that don't have anything to return, though?

Reply:
Oops, regardless sgtm